api/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -0,0 +1,309 @@
+	// If this field is present and contains at least one item, this rule
+	// allows/denies/passes traffic to the defined AdminNetworkPolicyPeer(s)
+	// If it is empty no traffic is matched by the AdminNetworkPolicyEgressRule.
+	To []AdminNetworkPolicyPeer `json:"to"`
+}
+
+// AdminNetworkPolicyPorts handles selection of traffic based on port(s).
+// Exactly one of the fields must be defined.
+type AdminNetworkPolicyPorts struct {
+	// AllPorts cannot be "false" when it is set
+	// AllPorts allows the user to select all ports for all protocols, thus not
+	// selecting traffic based on L4 principles.
+	// If "true" then all ports are selected for all protocols.
+	// +optional
+	AllPorts *bool `json:"allPorts,omitempty"`
+
+	// The list of ports to allow/deny/pass traffic on, each item in this list is
+	// combined using a logical OR. When this field is present it should contain at
+	// least one item, and this rule allows/denies/passes traffic only if the traffic
+	// matches at least one port in the list.
+	// +optional
+	List []AdminNetworkPolicyPort `json:"list,omitempty"`
+}
+
+// AdminNetworkPolicyPort describes a port to select
+type AdminNetworkPolicyPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+	// field defaults to TCP.
+	// +optional
+	// +kubebuilder:default=TCP
+	Protocol *v1.Protocol `json:"protocol,omitempty"`
+
+	// The port on the given protocol. This can either be a numerical or named
+	// port on a pod. If this field is not provided, this matches no port names and
+	// numbers.
+	// If present, only traffic on the specified protocol AND port will be matched.
+	// +optional
+	Port *intstr.IntOrString `json:"port,omitempty"`
+
+	// If set, indicates that the range of ports from port to endPort, inclusive,
+	// should be allowed by the policy. This field cannot be defined if the port field
+	// is not defined or if the port field is defined as a named (string) port.
+	// The endPort must be equal or greater than port.
+	// +optional
+	EndPort *int32 `json:"endPort,omitempty"`
+}
+
+// AdminNetPolRuleAction string describes the AdminNetworkPolicy action type.
+// +enum
+type AdminNetPolRuleAction string
+
+const (
+	// RuleActionPass enables admins to provide exceptions to ClusterNetworkPolicies and delegate this rule to
+	// K8s NetworkPolicies.
+	AdminNetPolRuleActionPass AdminNetPolRuleAction = "Pass"
+	// RuleActionDeny enables admins to deny specific traffic.
+	AdminNetPolRuleActionDeny AdminNetPolRuleAction = "Deny"
+	// RuleActionAllow enables admins to specifically allow certain traffic.
+	AdminNetPolRuleActionAllow AdminNetPolRuleAction = "Allow"
+)
+
+// AdminNetworkPolicyPeer defines an in-cluster peer to allow traffic to/from.
+// Exactly one of the selector pointers should be set for a given peer. If a
+// consumer observes none of its fields are set, they should assume an unknown
+// option has been specified and fail closed.
```

Comment:
"Fail closed" sounds nice, but think about what would actually happen.

If the rule has `action: Allow` and you don't recognize a peer, then you just ignore that peer. Some traffic that should have been allowed doesn't get allowed. OK.

If the rule has `action: Deny` and you don't recognize a peer, then you have to assume the worst case: the peer matches _all_ traffic. Everything that reaches this rule gets denied. Um.

If the rule has `action: Pass` and you don't recognize a peer, then... you have no idea if you're supposed to process further ANPs for this packet or not. So you probably need to treat it as blanket Deny again just to be safe.

In general, if the policy implementation is unable to correctly interpret an ANP, your cluster is probably going to explode.

Reply:
> In general, if the policy implementation is unable to correctly interpret an ANP, your cluster is probably going to explode.

See also the discussion about Subject above. ANP implementations cannot try to do something sane with a partially-understood ANP. I think we need to require the ANP implementation to provide an admission hook that rejects any ANP it cannot implement. (Though, even then we still have to worry about the "user trying to use features not present in the version of the CRD the apiserver knows about" problem.)

Reply:
Agreed, if we try to fail safely for every permutation of higher level ANP constructs (actions, ports etc) this becomes a mess... However, what if we just say, "If an unknown option has been specified in any `AdminNetworkPolicyPeer` then the ANP as a whole is broken, i.e block ALL traffic to the selected workloads, and report some sort of message in the status to prompt to user to fix things. " Then it will be easily evident that something is wrong to the user, and technically "safe" .

Reply:
> However, what if we just say, "If an unknown option has been specified in any AdminNetworkPolicyPeer then the ANP as a whole is broken, i.e block ALL traffic to the selected workloads, and report some sort of message in the status to prompt to user to fix things.

Compare:
- Scenario 1: Admin tries to install a third-party component in the cluster. The installation fails because the third-party component tries to install an ANP that uses an unsupported feature.
- Scenario 2: Admin tries to install a third-party component in the cluster. The installation succeeds. Later, a bunch users file urgent support tickets because their applications are suddenly completely broken and unreachable.

> then the ANP as a whole is broken, i.e block ALL traffic to the selected workloads

That assumes the implementation can identify "the selected workloads" even though it admittedly doesn't understand the entire ANP, and also assumes that we won't extend ANP in the future in ways that would expand the definition of "ALL traffic". (eg, it implies ANP could not be extended to support Multus interfaces in the future).

Reply:
Following up from the call...

In NetworkPolicy, since all rules just _allow_ traffic, we've generally felt like it's acceptable to have implementations "fail closed" by just not noticing / not understanding new features, and failing to allow the additional traffic allowed by the new features. (eg, An implementation that doesn't implement `endPort` will treat "allow ports 6000-6100" as if it was just "allow port 6000", so it's just failing to allow good traffic, not failing to deny bad traffic.)

In AdminNetworkPolicy, this does not work, because rules can have both "allow" and "deny" portions, so if an implementation ignores the parts it doesn't understand, it may end up failing to deny traffic. There's no sane way to try to partially implement an ANP with unrecognized bits.

So, how do we handle versioning / unimplemented features?

#### 1. How can an admin be sure, when creating a policy, that the policy only uses features that are part of the version of the CRD in this cluster?

(NP example: how can I know that the apiserver supports `endPort`?)

This is really not an ANP/NP-specific problem; _all_ types silently discard unrecognized fields, so trying to use a too-new feature of _any_ kube object may silently fail. I think the answer here is "tell users to **always use `kubectl create --validate=true`** or the equivalent if they aren't sure what the installed CRD supports".

#### 2. How can an ANP implementation be sure that an ANP in etcd was validated against the same version of the CRD that the implementation uses (and therefore doesn't have fields that the implementation can't see)?

(NP example: a network plugin built against `"k8s.io/api"` version 1.20 will not have the `EndPort` field in its `NetworkPolicyPort` struct, so when it receives `NetworkPolicy` objects from the apiserver, it will not be able to see port ranges and won't even know that they were there.)

I think the answer here is that implementations should **use the dynamic/untyped APIs** to get/list/watch ANPs, and then they can validate that decoding the `Unstructured` object into their own idea of the `AdminNetworkPolicy` object and then encoding back to `Unstructured` yields the same object they started with. If it doesn't, then that means the original object had some fields that aren't in the version of the golang struct that the ANP implementation is compiled against. (We can provide code somewhere, somehow, to help implementations out with this... Plugins may want to do this for NP too, though we don't currently have any good way for them to report back to the user if the user does try to create a "bad" NP. But maybe this solves some of the NP "Status" problems and we should resurrect that...)

#### 3. How can an administrator be sure, when creating a policy, that the policy only uses features that the network plugin implements?

From 1 we can say that the admin is responsible for only using features that the apiserver knows about, and from 2 we can say that the network plugin can tell whether a policy uses features that are newer than the network plugin. Additionally, it is reasonable to require network plugins to _recognize_ all of the features that are available in the version of the API that they were compiled against (even if they don't necessarily _implement_ all of them). Thus, we can say that when the plugin sees a given ANP, it should be able to determine whether it can implement it correctly or not. The question is: what happens next?

The PR currently suggests that if an implementation sees a policy it can't implement, it should set the policy's `status.conditions` appropriately. I argue that that's not very user-friendly, and instead **we should require the implementation to include an admission controller that just rejects any ANP that it cannot implement** (including in particular any ANP that uses features understood by the apiserver but not by the implementation).

(Note that this is different from the previous "admission controller" idea that was discussed; this is not an admission controller that we would write and then require administrators to run. It is an admission controller that would be part of the network plugin, written by the network plugin authors, installed as part of the network plugin.)

(Also note that the "how do I know the apiserver supports this object" part cannot be solved with an admission controller, because the object will have already been parsed against the apiserver's CRD before being passed to the admission controller. There was [discussion in the past about allowing CRDs to specify that they allow unrecognized properties](https://github.com/kubernetes/kubernetes/pull/58746) (which would then let us validate in the admission controller that no unrecognized properties were present), but the idea was rejected because it would cause problems. (eg, existing objects might fail to validate after upgrading to a new version of kube).)